### DISCUSSION TITLE
Fw.Time for TM/TC

### DISCUSSION THREAD

**Discussion by SMorettini** [nasa/fprime#2152] [Q&A]
Hi everyone,

I'm currently working on a project where I need to create a telemetry channel with a struct that contains an Fw.Time variable as data. The challenge I'm facing is that Fw.Time is not defined in any Fpp but only in a serializable, which means there is no direct definition of the type Fw::Time in the dictionary.

As of now, the workaround I've implemented involves using a custom Time type. Whenever I need to send data involving Fw.Time in TM/TC, I convert Fw.Time to my custom time type. However, I feel that this solution is not the most elegant.

Therefore, I'm interested in exploring better alternatives to handle this situation effectively. If there is a better way I would be happy to contribute to Fprime in the implementation.  

Thank you in advance for your assistance!

**Comment by LeStarch** [comment]
I think this should be tracked as an issue.  Note:  for telemetry channel, you can "override" the telemetry's time as an optional argument to the call, if the use case for sending time is to record the "real" sample time.

**Reply by SMorettini** [reply]
I'm not sure I understood correctly your note. 
Let's consider the struct:
```
  struct SomeData {
    my_data_1: F32
    Timestamp_data_1: Fw.Time
    my_data_2: F32
    timestamp_data_2: Fw.Time
  } 
``` 
I can pass this struct between components using ports without problem but when I want to send it to the ground station with a telemetry channel I have to convert it into something like:
```
  struct SomeData_ForTelemetry {
    my_data_1: F32
    Timestamp_data_1_seconds: U32
    Timestamp_data_1_Useconds: U32
    my_data_2: F32
    Timestamp_data_2_seconds: U32
    Timestamp_data_2_Useconds: U32
  } 
``` 
I think it would have been nicer if I could just use the first struct for internal communication and telemetry.
What do you mean by "override"?

**Reply by LeStarch** [reply]
It would not work in this case because you have combined 2 data entries with separate time stamps.  However, when you send telemetry, you can set the time:

```
Fw::Time event_occured_at = // some time
this->tlmWrite_CHANNEL(value, event_occured_at);
```
This allows you to "override" the internal timestamp on the channel to a more precise time rather than the moment the function was called.  This is often used when:

1. External time matches telemetry (e.g. time capturing register paired with value register)
2. Processing is done between telemetry capture and reporting, but accurate time is essential
3. Keeping several telemetry channels in-sync via shared time of capture.

Again, where you have multiple timestamps in the same struct this does not help as there is only one channel time.  I've marked this as an issue, so we can make your above use case possible.
